diskmaker/controllers/deleter/reconcile.go

```diff
@@ -123,29 +95,38 @@ type DeleteReconciler struct {
 	firstRunOver   bool
 }
 
-func (r *DeleteReconciler) SetupWithManager(mgr ctrl.Manager, cleanupTracker *provDeleter.CleanupStatusTracker, pvCache *provCache.VolumeCache) error {
+func NewDeleteReconciler(cleanupTracker *provDeleter.CleanupStatusTracker, pvCache *provCache.VolumeCache) *DeleteReconciler {
 
 	clientSet := provCommon.SetupClient()
 	runtimeConfig := &provCommon.RuntimeConfig{
 		UserConfig: &provCommon.UserConfig{
 			Node: &corev1.Node{},
 		},
-		Cache:    pvCache,
-		VolUtil:  provUtil.NewVolumeUtil(),
-		APIUtil:  provUtil.NewAPIUtil(clientSet),
-		Client:   clientSet,
-		Recorder: mgr.GetEventRecorderFor(ComponentName),
-		Mounter:  mount.New("" /* defaults to /bin/mount */),
-		// InformerFactory: , // unused
-
+		Cache:   pvCache,
+		VolUtil: provUtil.NewVolumeUtil(),
+		APIUtil: provUtil.NewAPIUtil(clientSet),
+		Client:  clientSet,
+		Mounter: mount.New("" /* defaults to /bin/mount */),
 	}
 
-	r.runtimeConfig = runtimeConfig
-	r.deleter = &provDeleter.Deleter{
+	deleter := &provDeleter.Deleter{
 		RuntimeConfig: runtimeConfig,
 		CleanupStatus: cleanupTracker,
 	}
-	return ctrl.NewControllerManagedBy(mgr).
+
+	deleteReconciler := &DeleteReconciler{
+		runtimeConfig: runtimeConfig,
+		deleter:       deleter,
+	}
+
+	return deleteReconciler
+}
+
+func (r *DeleteReconciler) WithManager(mgr ctrl.Manager) (*DeleteReconciler, error) {
+
+	r.runtimeConfig.Recorder = mgr.GetEventRecorderFor(ComponentName)
```

Comment:
Changing runtimeConfig on multiple places like this is not ideal, however I'm not sure what change would improve it without changing the structure too much. The problem is that the runtimeConfig contains many values that are available at different parts of the flow, most of them are obtained only after Reconcile() starts.

We could remove WithManager and make ctrl.Manager part of the constructor to make it cleaner, but I'm not sure if there might be a case where manager is not available.

More ideas are welcome.